app/models/binary/ThumbnailService.scala

```diff
@@ -0,0 +1,237 @@
+package models.binary
+import com.scalableminds.util.accesscontext.{DBAccessContext, GlobalAccessContext}
+import com.scalableminds.util.cache.AlfuCache
+import com.scalableminds.util.geometry.{BoundingBox, Vec3Int}
+import com.scalableminds.util.mvc.MimeTypes
+import com.scalableminds.util.time.Instant
+import com.scalableminds.util.tools.Fox.option2Fox
+import com.scalableminds.util.tools.{Fox, JsonHelper}
+import com.scalableminds.webknossos.datastore.models.datasource.DataSetViewConfiguration.DataSetViewConfiguration
+import com.scalableminds.webknossos.datastore.models.datasource.{DataLayerLike, GenericDataSource}
+import com.typesafe.scalalogging.LazyLogging
+import models.configuration.DataSetConfigurationService
+import net.liftweb.common.Full
+import play.api.http.Status.NOT_FOUND
+import play.api.i18n.{Messages, MessagesProvider}
+import play.api.libs.json.JsArray
+import utils.ObjectId
+import utils.sql.{SimpleSQLDAO, SqlClient}
+
+import javax.inject.Inject
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+class ThumbnailService @Inject()(dataSetService: DataSetService,
+                                 thumbnailCachingService: ThumbnailCachingService,
+                                 dataSetConfigurationService: DataSetConfigurationService,
+                                 dataSetDAO: DataSetDAO,
+                                 thumbnailDAO: ThumbnailDAO)
+    extends LazyLogging
+    with MimeTypes {
+
+  private val DefaultThumbnailWidth = 400
+  private val DefaultThumbnailHeight = 400
+  private val MaxThumbnailWidth = 4000
+  private val MaxThumbnailHeight = 4000
+
+  def getThumbnailWithCache(
+      organizationName: String,
+      datasetName: String,
+      layerName: String,
+      w: Option[Int],
+      h: Option[Int],
+      mappingName: Option[String])(implicit ec: ExecutionContext, mp: MessagesProvider): Fox[Array[Byte]] = {
+    val width = com.scalableminds.util.tools.Math.clamp(w.getOrElse(DefaultThumbnailWidth), 1, MaxThumbnailWidth)
```

Comment:
Interesting that this clamp does not use the commonly used order of min, value, max which is also used in frontend utils.ts and common library functions. Made me stumble here.

Reply:
I googled around a bit and it looks like the ordering `value, min, max` is actually also pretty common across languages (e.g. [c++](https://en.cppreference.com/w/cpp/algorithm/clamp) or [c#](https://www.educative.io/answers/what-is-mathclamp-in-c-sharp), [unity3d library](https://docs.unity3d.com/ScriptReference/Mathf.Clamp.html)). On stackoverflow it’s about 50/50, compare [here](https://stackoverflow.com/questions/16656651/does-java-have-a-clamp-function).

I think I’d leave this as is for now unless you would vote otherwise?

Reply:
No strong opinion here, just thought it weird